### Fitting Logistic Regression

In this first notebook, you will be fitting a logistic regression model to a dataset where we would like to predict if a transaction is fraud or not.

To get started let's read in the libraries and take a quick look at the dataset.

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm


df = pd.read_csv('./fraud_dataset.csv')
df.head()

/opt/conda/lib/python3.6/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


transaction_id   duration      day  fraud
0           28891  21.302600  weekend  False
1           61629  22.932765  weekend  False
2           53707  32.694992  weekday  False
3           47812  32.784252  weekend  False
4           43455  17.756828  weekend  False

`1.` As you can see, there are two columns that need to be changed to dummy variables.  Replace each of the current columns to the dummy version.  Use the 1 for `weekday` and `True`, and 0 otherwise.  Use the first quiz to answer a few questions about the dataset.

In [4]:
df['fraud'] = pd.get_dummies(df['fraud'], drop_first = True)

In [9]:
df['day'] = pd.get_dummies(df['day'])['weekday']

In [10]:
df.head()

transaction_id   duration  day  fraud
0           28891  21.302600    0      0
1           61629  22.932765    0      0
2           53707  32.694992    1      0
3           47812  32.784252    0      0
4           43455  17.756828    0      0

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8793 entries, 0 to 8792
Data columns (total 5 columns):
transaction_id    8793 non-null int64
duration          8793 non-null float64
day               8793 non-null uint8
fraud             8793 non-null uint8
intercept         8793 non-null int64
dtypes: float64(1), int64(2), uint8(2)
memory usage: 223.3 KB


In [16]:
1 - df['fraud'].mean()

0.9878312293870124

In [17]:
df[df['fraud'] == 1]['duration'].mean()

4.6242473706156568

In [20]:
df['day'].mean()

0.34527465029000343

In [21]:
df[df['fraud'] == 0]['duration'].mean()

30.013583132522555

`2.` Now that you have dummy variables, fit a logistic regression model to predict if a transaction is fraud using both day and duration.  Don't forget an intercept!  Use the second quiz below to assure you fit the model correctly.

In [22]:
df['intercept'] = 1
model = sm.Logit(df['fraud'], df[['intercept', 'duration', 'day']])
results = model.fit()
results.summary()

Optimization terminated successfully.
         Current function value: inf
         Iterations 16


/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1214: RuntimeWarning: overflow encountered in exp
  return 1/(1+np.exp(-X))
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:1264: RuntimeWarning: divide by zero encountered in log
  return np.sum(np.log(self.cdf(q*np.dot(X,params))))
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:2397: RuntimeWarning: invalid value encountered in double_scalars
  return 1 - self.llf/self.llnull
/opt/conda/lib/python3.6/site-packages/statsmodels/discrete/discrete_model.py:2401: RuntimeWarning: invalid value encountered in double_scalars
  return -2*(self.llnull - self.llf)


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  fraud   No. Observations:                 8793
Model:                          Logit   Df Residuals:                     8790
Method:                           MLE   Df Model:                            2
Date:                Sat, 02 Mar 2019   Pseudo R-squ.:                     nan
Time:                        20:06:09   Log-Likelihood:                   -inf
converged:                       True   LL-Null:                          -inf
                                        LLR p-value:                       nan
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
intercept      9.8709      1.944      5.078      0.000       6.061      13.681
duration      -1.4637      0.290     -5.039      0.000      -2.033      -0.894
day            2.5465      0.904      2.816      0.005       0.774       4.319
==============================================================================

Possibly complete quasi-separation: A fraction 0.98 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""


The proportion of fraudulent transactions = 0.012

The average duration for fraudulent transaction = 4.62


The proportion of weekday transactions = 0.345


The average duration for non-fraudulent transactions = 30.01


Which variables are statistically significant for predicting fraud Duration and Weekday.

Following interpretations of data: 

On weekdays, the chance of fraud is 12.76 times more likely than on weekends holding duration constant.

For each minute less spent on the transaction, the chance of fraud is 4.32 times more likely holding the day of week constant.